aptos-move/aptos-vm/src/sharded_block_executor/cross_shard_client.rs

```diff
@@ -0,0 +1,160 @@
+                    let (state_key, write_op) = txn_commit_msg.take();
+                    cross_shard_state_view
+                        .set_value(&state_key, write_op.and_then(|w| w.as_state_value()));
+                },
+                CrossShardMsg::StopMsg => {
+                    break;
+                },
+            }
+        }
+    }
+}
+
+pub struct CrossShardCommitSender {
+    shard_id: ShardId,
+    // The senders of cross-shard messages to other shards.
+    message_txs: Vec<Mutex<Sender<CrossShardMsg>>>,
+    // The hashmap of source txn index to hashmap of conflicting storage location to the
+    // list of target txn index and shard id. Please note that the transaction indices stored here is
+    // global indices, so we need to convert the local index received from the parallel execution to
+    // the global index.
+    dependent_edges: HashMap<TxnIndex, HashMap<StateKey, HashSet<ShardId>>>,
+    // The offset of the first transaction in the sub-block. This is used to convert the local index
+    // in parallel execution to the global index.
+    index_offset: TxnIndex,
+}
+
+impl CrossShardCommitSender {
+    pub fn new(
+        shard_id: ShardId,
+        message_txs: Vec<Sender<CrossShardMsg>>,
+        sub_block: &SubBlock<Transaction>,
+    ) -> Self {
+        let mut dependent_edges = HashMap::new();
+        let mut num_dependent_edges = 0;
+        for (txn_idx, txn_with_deps) in sub_block.txn_with_index_iter() {
+            let mut storage_locations_to_target = HashMap::new();
+            for (txn_id_with_shard, storage_locations) in txn_with_deps
+                .cross_shard_dependencies
+                .dependent_edges()
+                .iter()
+            {
+                for storage_location in storage_locations {
+                    storage_locations_to_target
+                        .entry(storage_location.clone().into_state_key())
+                        .or_insert_with(HashSet::new)
+                        .insert(txn_id_with_shard.shard_id);
+                    num_dependent_edges += 1;
+                }
+            }
+            if !storage_locations_to_target.is_empty() {
+                dependent_edges.insert(txn_idx as TxnIndex, storage_locations_to_target);
+            }
+        }
+
+        trace!(
+            "CrossShardCommitSender::new: shard_id: {:?}, num_dependent_edges: {:?}",
+            shard_id,
+            num_dependent_edges
+        );
+
+        Self {
+            shard_id,
+            message_txs: message_txs.into_iter().map(Mutex::new).collect(),
+            dependent_edges,
+            index_offset: sub_block.start_index as TxnIndex,
+        }
+    }
+
+    fn send_remote_update_for_success(
+        &self,
+        txn_idx: TxnIndex,
+        txn_output: &AptosTransactionOutput,
+    ) {
+        let edges = self.dependent_edges.get(&txn_idx).unwrap();
+        let write_set = txn_output.committed_output().unwrap().write_set();
```

Comment:
This unwrap failed in my sequential execution testing: inside the `txn_output`, the `VmOutput` hasn't been converted to `TransactionOutput`.

I had to do a workaround in my testing branch to do the conversion explicitly like [this](https://github.com/aptos-labs/aptos-core/pull/8852/commits/2ae60dbf593c92ffcb165e9182f9b747482abb95#r1247026775). See the full commit for the workaround [here](https://github.com/aptos-labs/aptos-core/pull/8852/commits/2ae60dbf593c92ffcb165e9182f9b747482abb95).

Reply:
Thanks for looking into this. Let me address this in a separate follow up PR.